src/main/java/com/ss/training/utopia/counter/service/BookingService.java

```diff
@@ -0,0 +1,96 @@
+package com.ss.training.utopia.counter.service;
+
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import com.ss.training.utopia.counter.dao.AirportDao;
+import com.ss.training.utopia.counter.dao.BookingDao;
+import com.ss.training.utopia.counter.dao.FlightDao;
+import com.ss.training.utopia.counter.dao.UserDao;
+import com.ss.training.utopia.counter.entity.Airport;
+import com.ss.training.utopia.counter.entity.Booking;
+import com.ss.training.utopia.counter.entity.Flight;
+import com.ss.training.utopia.counter.entity.User;
+
+/**
+ * @author Justin O'Brien
+ */
+@Service
+public class BookingService {
+
+	@Autowired
+	AirportDao airportDao;
+	@Autowired
+	UserDao userDao;
+	@Autowired
+	FlightDao flightDao;
+	@Autowired
+	BookingDao bookingDao;
+
+	public Boolean createUser(User user) {
+		user.setPassword(new BCryptPasswordEncoder().encode(user.getPassword()));
+		try {
+			userDao.save(user);
+		} catch (Throwable t) {
+			return null;
+		}
+		return true;
+	}
+
+	public Boolean usernameAvailable(String username) {
+		User user;
+		try {
+			user = userDao.findByUsername(username);
+		} catch (Throwable t) {
+			return null;
+		}
+		return user == null;
+	}
+
+	public Airport[] getAllAirports() {
+		List<Airport> airports;
+		try {
+			airports = airportDao.findAll();
+		} catch (Throwable t) {
+			return null;
+		}
+		return airports.toArray(new Airport[airports.size()]);
+	}
+
+	public Flight[] getBookableFlights(Long departId, Long arriveId, Long travelerId) {
+		List<Flight> flights;
+		try {
+			flights = flightDao.findBookable(departId, arriveId, travelerId);
+		} catch (Throwable t) {
+			return null;
+		}
+		return flights.toArray(new Flight[flights.size()]);
+	}
+
+	@Transactional
+	public Boolean bookFlight(Booking booking) {
+		Flight flight;
+		try {
+			flight = flightDao.findByFlightId(booking.getFlightId());
+		} catch (Throwable t) {
+			return null;
+		}
+		if (flight.getSeatsAvailable() <= 0)
```

Comment:
This is good to check. I was thinking that it would not be possible for people to book a flight with no tickets because they wouldn't see it, but this accounts for a bunch of people all competing for a last ticket and making the call at the same time.

Reply:
Yep, or if the user just stays on the page that lists available flights for a while and by the time they pick one it's not available anymore.